### PR TITLE
Use ActiveJob 5.2 retry logic for old jobs

### LINKED ISSUE #35986: `ActiveJob` `exception_executions` may cause incorrect / unexpected retry behavior during upgrade
I'm opening this in hopes of getting some feedback and opinions about a potential issue sites may encounter during upgrade. I have a "soft" opinion, but no preferred hard-and-fast solution. I'm also very open to someone pointing out that my understanding is wrong and that this is a non-issue entirely.

**TL/DR;**

* I'm hoping to get some feedback about whether this is really an issue and what is the best thing to be done about it if it is.

* `ActiveJob` execution counts and `retry_on` logic is handled differently for Rails 5 and Rails 6. This can potentially pose problems in a downgraded or temporarily mixed production environment.

* During testing with production data or final rollout to production, Rails 5 and 6 applications can fight over what counter data should be serialized for the job and how to handle retries. This can happen if the app is production tested with Rails 6 and then rolled back to Rails 5 or when Rails 6 is deployed to production and Rails 5 workers have not yet finished deploying.

* It may be better for Rails 6.0 to support both the rails 5 retry method as well as Rails 6 and allow the application to opt in when Rails 6 workers are deployed and ready to handle the new retry method.

**PR History**

1. Last November: https://github.com/rails/rails/pull/34352 added `exception_executions` to `ActiveJob`and allowed `ActiveJob` to correctly track execution count based on the exceptions passed to `retry_on`. (Fixed https://github.com/rails/rails/issues/34337)

2. Last January: https://github.com/rails/rails/pull/34731 fixed an issue in which Rails 6 workers processing jobs queued by Rails 5 would error because `exception_executions` was `nil`. The fix here was to initialize `exception_executions` as needed if it was found to be missing.

3. Last Week (April): https://github.com/rails/rails/pull/35956 addressed calculating the correct delay time for retries. This changed the delay to be based on `exception_executions` instead of the global `executions` count.

**The Nature of the Problem**

The November PR is a fix for what DHH noted was a bug (https://github.com/rails/rails/issues/34337#issuecomment-434004236), however current production applications still expect the current (buggy) retry behavior. The process of migrating from Rails 5 to 6 with jobs in the production queue has the potential to be problematic and produce unexpected results with retries.

A couple of issues to consider:

* `ActiveJob` instances queued by Rails 6 applications need to be reliably run by Rails 5 worker applications and vice versa.

* When testing for Rails 6 begins in production, the Rails 6 application will be touching production data. The way it stands now, Rails 5 and Rails 6 serialization methods will be fighting over whether or not `exception_executions` is part of the serialized data, and whether or not to keep retrying the job based on one of two different execution counts.

* It will be impossible to deploy to all production servers at exactly the same time, so there will necessarily be some amount of time in which the main application and the background workers are on two different versions. If testing is done in a way where Rails 6 is rolled out, tested, and reverted to Rails 5, then the mismatch can be in either direction.

**Based on the current behavior, it seems to me that the following has potential to occur:**

* A job enqueued with Rails 5 will have a global `executions` counter and not a `exception_executions` counter.

* If Rails 6 tries to perform the job, the counter will be found missing, initialized (set to zero), and not fail. However, previous execution counts will be ignored, such that if the job failed `2` times already, it will now have an exception-specific execution count of `0`.

* On the next failure, the Rails 6 worker will save updates to both the `executions` and `exception_executions` counts. (the global count functionality is not lost)

* If a Rails 5 worker then tries to run the job, such as after rolling back from a production test, it will do so using the global `executions` counter, which may be higher than normal because Rails 6 allows multiple exceptions to reach their own maximum number but still increments the global execution count. This will revert to the Rails 5 behavior and, if the global counter is > the max exceptions allowed, it will stop retrying the job. (This is probably expected and reasonable behavior.)

* Jobs enqueued with Rails 6, or run by a Rails 6 worker (e.g. during production testing) will have both `executions` and `exception_executions`. Rails 5 workers that subsequently run the job will remove the `exception_executions` key from the serialized data, effectively turning it back into a Rails 5 job. If a Rails 6 worker then picks it up.... (repeat)

**I believe the root of the concern may be that:**

1. When upgrading to rails 6, the new retry behavior is immediately enforced, without giving any time to allow existing queued jobs to finish being filtered through the pipeline or for Rails 6 upgrade deploys to fully finish rolling out to all servers / workers.

2. When Rails 6 encounters a job enqueued by Rails 5, it attempts to update it's `exception_executions` value and make the job Rails 6 compatible instead of continuing to run the job "the Rails 5 way".

I think there may be a case to say that:

* Rails 6.0 retry behavior should accommodate both global and per-exception counts and be executed in the way expected at the time the job was added to the queue instead of based on the version of Rails running the job.

* Rails 6.0 jobs should default to the old execution behavior when adding to the queue and allow the application to opt into the new behavior when there are Rails 6 workers ready to handle the newly queued jobs. Or at least, allow the application to opt out of the new behavior until ready.

* Rails the 5 retry process should be deprecated in Rails 6.0 and removed in Rails 6.1.

I'd really love to hear additional opinions on this and whether or not the original PR authors have any thoughts on handling the upgrade transition in a more graceful way.

CC PR Authors: @albertoalmagro, @rosa, @mrhead

### PR DESCRIPTION
**Pull request by jhawthorn**
Rails 6 introduces retries per-exception, instead of a global count of retries. Because ActiveJob 5.2 doesn't serialize the execution count per-exception, when ActiveJob 6.0 picks up an "old" job it can't know the exception count in the new format.

This can also be an issue if AJ 6.0 serializes a new job with exception_executions which is later picked up by AJ 5.2, which would clear exception_executions (since it has no knowledge of it).

Previously we handled this by resetting exception_executions, if it wasn't defined on a job, which could result in the worst case retrying the job 2x the times we should.

This commit changes how we handle loading a legacy job: instead of resetting exception_executions, we instead will always use the global executions count.

This way, jobs which only have one retry_on (and didn't have a behaviour change in AJ 6) are backwards-and-forwards-compatible with counts respected exactly.

Jobs with multiple retry_on will revert to the AJ5.2 behaviour if they were ever run under AJ5.2.

Fixes #35986 

cc @rosa @cpruitt @kaspth @tenderlove @eileencodes 🙇

### CODE REVIEW

File: activejob/lib/active_job/exceptions.rb [hunk 1]
```diff
@@ -146,5 +144,15 @@ def instrument(name, error: nil, wait: nil, &block)
 
         ActiveSupport::Notifications.instrument("#{name}.active_job", payload, &block)
       end
+
+      def executions_for(exceptions)
+        if self.exception_executions
```
Comment (kaspth) [inline activejob/lib/active_job/exceptions.rb]: Can spare `self.` here.

File: activejob/lib/active_job/exceptions.rb [hunk 2]
```diff
@@ -146,5 +144,15 @@ def instrument(name, error: nil, wait: nil, &block)
...
+          self.exception_executions[exceptions.to_s] = (exception_executions[exceptions.to_s] || 0) + 1
+        else
+          # Guard against jobs that were persisted before we started having individual executions counters per retry_on
+          self.executions
```
Comment (kaspth) [inline activejob/lib/active_job/exceptions.rb]: Can spare `self.` here.

**Review by kaspth — COMMENTED** [review]
Is it possible to write a test job where we mutate the arguments after serialization, maybe?

Ideally we'd have a 5.2 job that we process on 6.0 and the other way around.

Perhaps something like the tests here is possible? https://github.com/rails/rails/pull/26017/files

File: activejob/lib/active_job/exceptions.rb [hunk 3]
```diff
@@ -146,5 +144,15 @@ def instrument(name, error: nil, wait: nil, &block)
...
+
+      def executions_for(exceptions)
+        if exception_executions
+          exception_executions[exceptions.to_s] = (exception_executions[exceptions.to_s] || 0) + 1
+        else
+          # Guard against jobs that were persisted before we started having individual executions counters per retry_on
+          executions
+        end
+      end
+
```
Comment (rafaelfranca) [inline activejob/lib/active_job/exceptions.rb]: ✂️

**Comment by rafaelfranca** [comment]
Review @rosa

**Review by rafaelfranca — APPROVED** [review]

**Comment by jhawthorn** [comment]
Rebased and added tests.

**Review by kaspth — APPROVED** [review]
Dig the extra test coverage!

I guess the main way to conceptualize this is that if Active Job 5.2 touches a job at any point `exception_executions` will be wiped on re-enqueue. So a job with max 5 retries enqueued on 6.0 and attempted 4 times on 6.0, if it then got picked up by 5.2 it would only see `executions: 0`. The job would then execute up to 2x of the max in the worst case as you mentioned.

**Review by rosa — APPROVED** [review]
Looks great, I really like this simple approach to support the transition 💯Thanks @jhawthorn!

**Comment by rosa** [comment]
@kaspth I think it's the other way around, from 5.2 to 6.0, because in 6.0 we still increment `executions`. A job enqueued on 6.0 and attempted 4 times on 6.0 would have `executions == 4` when getting picked by 5.2, and will be retried the right number of times. However, if it starts in 5.2, and it's retried 4 times, and then picked up by 6.0, it'd have `exception_executions == nil`, defaulting to 0 and getting retried again (up to ~2x in the worst case as you said).

**Comment by kaspth** [comment]
@rosa Oh! I hadn't picked up that Rails 6 still incremented `executions`. But of course it does for backwards compatibility. Thanks for the explanation!

**Pull request merged**